File: sdks/node/src/openfeature.ts
/**
 * OpenFeature provider for FeatureSignals Node.js SDK.
 *
 * Usage:
 * ```ts
 * import { FeatureSignalsClient } from "@featuresignals/node";
 * import { FeatureSignalsProvider } from "@featuresignals/node/openfeature";
 * import { OpenFeature } from "@openfeature/server-sdk";
 *
 * const fsClient = new FeatureSignalsClient("fs_srv_...", { envKey: "production" });
 * await fsClient.waitForReady();
 *
 * await OpenFeature.setProviderAndWait(new FeatureSignalsProvider(fsClient));
 * const client = OpenFeature.getClient();
 * const enabled = await client.getBooleanValue("dark-mode", false);
 * ```
 */

import type { FeatureSignalsClient } from "./client.ts";

export interface EvaluationContext {
  targetingKey?: string;
  [key: string]: unknown;
}

export interface ResolutionDetails<T> {
  value: T;
  reason?: string;
  errorCode?: string;
  errorMessage?: string;
}

export interface ProviderMetadata {
  name: string;
}

/**
 * FeatureSignalsProvider wraps a FeatureSignalsClient and exposes the
 * OpenFeature provider interface for server-side Node.js usage.
 *
 * All evaluations are local lookups against the client's cached flags.
 */
export class FeatureSignalsProvider {
  readonly metadata: ProviderMetadata = { name: "FeatureSignals" };
  private client: FeatureSignalsClient;

  constructor(client: FeatureSignalsClient) {
    this.client = client;
  }

  resolveBooleanEvaluation(
    flagKey: string,
    defaultValue: boolean,
    context?: EvaluationContext,
  ): ResolutionDetails<boolean> {
    const val = this.client.allFlags()[flagKey];
    if (val === undefined) {
      return {
        value: defaultValue,
        reason: "ERROR",
        errorCode: "FLAG_NOT_FOUND",
        errorMessage: `Flag '${flagKey}' not found`,
      };
    }
    if (typeof val !== "boolean") {
      return {
        value: defaultValue,
        reason: "ERROR",
        errorCode: "TYPE_MISMATCH",
        errorMessage: `Flag '${flagKey}' is not boolean`,
      };
    }
    return { value: val, reason: "CACHED" };
  }

  resolveStringEvaluation(
    flagKey: string,
    defaultValue: string,
    context?: EvaluationContext,
  ): ResolutionDetails<string> {
    const val = this.client.allFlags()[flagKey];
    if (val === undefined) {
      return {
        value: defaultValue,
        reason: "ERROR",
        errorCode: "FLAG_NOT_FOUND",
        errorMessage: `Flag '${flagKey}' not found`,
      };
    }
    if (typeof val !== "string") {
      return {
        value: defaultValue,
        reason: "ERROR",
        errorCode: "TYPE_MISMATCH",
        errorMessage: `Flag '${flagKey}' is not a string`,
      };
    }
    return { value: val, reason: "CACHED" };
  }

  resolveNumberEvaluation(
    flagKey: string,
    defaultValue: number,
    context?: EvaluationContext,
  ): ResolutionDetails<number> {
    const val = this.client.allFlags()[flagKey];
    if (val === undefined) {
      return {
        value: defaultValue,
        reason: "ERROR",
        errorCode: "FLAG_NOT_FOUND",
        errorMessage: `Flag '${flagKey}' not found`,
      };
    }
    if (typeof val !== "number") {
      return {
        value: defaultValue,
        reason: "ERROR",
        errorCode: "TYPE_MISMATCH",
        errorMessage: `Flag '${flagKey}' is not a number`,
      };
    }
    return { value: val, reason: "CACHED" };
  }

  resolveObjectEvaluation<T>(
    flagKey: string,
    defaultValue: T,
    context?: EvaluationContext,
  ): ResolutionDetails<T> {
    const val = this.client.allFlags()[flagKey];
    if (val === undefined) {
      return {
        value: defaultValue,
        reason: "ERROR",
        errorCode: "FLAG_NOT_FOUND",
        errorMessage: `Flag '${flagKey}' not found`,
      };
    }
    return { value: val as T, reason: "CACHED" };
  }
}

File: sdks/go/openfeature.go
package featuresignals

import (
	"context"

	of "github.com/open-feature/go-sdk/openfeature"
)

// Provider implements the OpenFeature FeatureProvider interface, backed by a
// FeatureSignals Client. All evaluations are local lookups against the client's
// cached flag values — no additional network calls are made.
//
// Usage:
//
//	client := featuresignals.NewClient("fs_srv_...", "production",
//	    featuresignals.WithBaseURL("http://localhost:8080"),
//	)
//	<-client.Ready()
//
//	of.SetProviderAndWait(featuresignals.NewProvider(client))
//	ofClient := of.NewClient("my-service")
//	enabled, _ := ofClient.BooleanValue(ctx, "dark-mode", false, of.EvaluationContext{})
type Provider struct {
	client *Client
}

var _ of.FeatureProvider = (*Provider)(nil)

// NewProvider creates an OpenFeature provider backed by the given Client.
func NewProvider(client *Client) *Provider {
	return &Provider{client: client}
}

func (p *Provider) Metadata() of.Metadata {
	return of.Metadata{Name: "FeatureSignals"}
}

func (p *Provider) Hooks() []of.Hook {
	return nil
}

func (p *Provider) BooleanEvaluation(_ context.Context, flag string, defaultValue bool, evalCtx of.FlattenedContext) of.BoolResolutionDetail {
	v, ok := p.client.getFlag(flag)
	if !ok {
		return of.BoolResolutionDetail{
			Value: defaultValue,
			ProviderResolutionDetail: of.ProviderResolutionDetail{
				ResolutionError: of.NewFlagNotFoundResolutionError(flag),
				Reason:          of.ErrorReason,
			},
		}
	}
	b, ok := v.(bool)
	if !ok {
		return of.BoolResolutionDetail{
			Value: defaultValue,
			ProviderResolutionDetail: of.ProviderResolutionDetail{
				ResolutionError: of.NewTypeMismatchResolutionError("expected bool"),
				Reason:          of.ErrorReason,
			},
		}
	}
	return of.BoolResolutionDetail{
		Value:                    b,
		ProviderResolutionDetail: of.ProviderResolutionDetail{Reason: of.CachedReason},
	}
}

func (p *Provider) StringEvaluation(_ context.Context, flag string, defaultValue string, evalCtx of.FlattenedContext) of.StringResolutionDetail {
	v, ok := p.client.getFlag(flag)
	if !ok {
		return of.StringResolutionDetail{
			Value: defaultValue,
			ProviderResolutionDetail: of.ProviderResolutionDetail{
				ResolutionError: of.NewFlagNotFoundResolutionError(flag),
				Reason:          of.ErrorReason,
			},
		}
	}
	s, ok := v.(string)
	if !ok {
		return of.StringResolutionDetail{
			Value: defaultValue,
			ProviderResolutionDetail: of.ProviderResolutionDetail{
				ResolutionError: of.NewTypeMismatchResolutionError("expected string"),
				Reason:          of.ErrorReason,
			},
		}
	}
	return of.StringResolutionDetail{
		Value:                    s,
		ProviderResolutionDetail: of.ProviderResolutionDetail{Reason: of.CachedReason},
	}
}

func (p *Provider) FloatEvaluation(_ context.Context, flag string, defaultValue float64, evalCtx of.FlattenedContext) of.FloatResolutionDetail {
	v, ok := p.client.getFlag(flag)
	if !ok {
		return of.FloatResolutionDetail{
			Value: defaultValue,
			ProviderResolutionDetail: of.ProviderResolutionDetail{
				ResolutionError: of.NewFlagNotFoundResolutionError(flag),
				Reason:          of.ErrorReason,
			},
		}
	}
	f, ok := v.(float64)
	if !ok {
		return of.FloatResolutionDetail{
			Value: defaultValue,
			ProviderResolutionDetail: of.ProviderResolutionDetail{
				ResolutionError: of.NewTypeMismatchResolutionError("expected float64"),
				Reason:          of.ErrorReason,
			},
		}
	}
	return of.FloatResolutionDetail{
		Value:                    f,
		ProviderResolutionDetail: of.ProviderResolutionDetail{Reason: of.CachedReason},
	}
}

func (p *Provider) IntEvaluation(_ context.Context, flag string, defaultValue int64, evalCtx of.FlattenedContext) of.IntResolutionDetail {
	v, ok := p.client.getFlag(flag)
	if !ok {
		return of.IntResolutionDetail{
			Value: defaultValue,
			ProviderResolutionDetail: of.ProviderResolutionDetail{
				ResolutionError: of.NewFlagNotFoundResolutionError(flag),
				Reason:          of.ErrorReason,
			},
		}
	}
	// JSON numbers unmarshal as float64 in Go
	f, ok := v.(float64)
	if !ok {
		return of.IntResolutionDetail{
			Value: defaultValue,
			ProviderResolutionDetail: of.ProviderResolutionDetail{
				ResolutionError: of.NewTypeMismatchResolutionError("expected numeric"),
				Reason:          of.ErrorReason,
			},
		}
	}
	return of.IntResolutionDetail{
		Value:                    int64(f),
		ProviderResolutionDetail: of.ProviderResolutionDetail{Reason: of.CachedReason},
	}
}

func (p *Provider) ObjectEvaluation(_ context.Context, flag string, defaultValue interface{}, evalCtx of.FlattenedContext) of.InterfaceResolutionDetail {
	v, ok := p.client.getFlag(flag)
	if !ok {
		return of.InterfaceResolutionDetail{
			Value: defaultValue,
			ProviderResolutionDetail: of.ProviderResolutionDetail{
				ResolutionError: of.NewFlagNotFoundResolutionError(flag),
				Reason:          of.ErrorReason,
			},
		}
	}
	return of.InterfaceResolutionDetail{
		Value:                    v,
		ProviderResolutionDetail: of.ProviderResolutionDetail{Reason: of.CachedReason},
	}
}
